webhook/escalation/webhook.go

```diff
@@ -0,0 +1,270 @@
+		r.Context(),
+		escalation.Spec.PolicyName,
+		metav1.GetOptions{},
+	)
+
+	switch {
+	case errors.IsNotFound(err):
+		klog.InfoS(
+			"User submitted an escalation request refering to a policy that doesn't exist",
+			usernameAndPolicyTags(
+				review.Request.UserInfo.Username,
+				"unknown",
+			)...,
+		)
+
+		review.Response.Result = &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: fmt.Sprintf("Unknown policy: %s", escalation.Spec.PolicyName),
+		}
+
+		webhooksupport.WriteJSON(rw, http.StatusOK, &review)
+		return
+	case err != nil:
+		klog.ErrorS(err, "Unable to retrieve escalation policy")
+		review.Response.Result = &UnexpectedErrorStatus
+
+		webhooksupport.WriteJSON(rw, http.StatusOK, &review)
+		return
+	default:
+		// We're good.
+	}
+
+	if !userAllowed(*policy, review.Request.UserInfo) {
+		klog.InfoS(
+			"User attempted to use an escalation policy, but is not part of the policy subjects",
+			usernameAndPolicyTags(
+				review.Request.UserInfo.Username,
+				policy.Name,
+			)...,
+		)
+
+		review.Response.Result = &metav1.Status{
+			Status: metav1.StatusFailure,
+			Message: fmt.Sprintf(
+				"User %q is not allowed to use the escalation policy %q",
+				review.Request.UserInfo.Username,
+				policy.Name,
+			),
+		}
+
+		webhooksupport.WriteJSON(rw, http.StatusOK, &review)
+		return
+	}
+
+	if review.Response.Patch, err = genObjectPatch(review.Request.UserInfo); err != nil {
+		klog.ErrorS(
+			err,
+			"Unable to generate object patch",
+			usernameAndPolicyTags(
+				review.Request.UserInfo.Username,
+				policy.Name,
+			)...,
+		)
+
+		review.Response.Result = &UnexpectedErrorStatus
+
+		webhooksupport.WriteJSON(rw, http.StatusOK, &review)
+		return
+
+	}
+
+	klog.InfoS(
+		"User submitted an escalation request",
+		"username",
+		review.Request.UserInfo.Username,
+		"policy",
+		escalation.Spec.PolicyName,
+	)
+
+	review.Response.Allowed = true
+	review.Response.Result = &metav1.Status{Status: metav1.StatusSuccess}
+	review.Response.PatchType = webhooksupport.Ptr(admissionv1.PatchTypeJSONPatch)
+
+	webhooksupport.WriteJSON(rw, http.StatusOK, &review)
+}
+
+// userAllowed returns true when an user is allowed to use an escalation policy based
+// on the policy subjects.
+// An user is allowed if and only if one of the policy subject:
+// - is of Kind user with the same username.
+// - is of Kind group and the user belongs to this group.
+func userAllowed(policy kudov1alpha1.EscalationPolicy, user authenticationv1.UserInfo) bool {
+	for _, subject := range policy.Spec.Subjects {
+		switch subject.Kind {
+		case rbacv1.GroupKind:
+			if userInGroup(user.Groups, subject.Name) {
+				return true
+			}
+		case rbacv1.UserKind:
+			if subject.Name == user.Username {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func userInGroup(userGroups []string, groupName string) bool {
+	for _, userGroup := range userGroups {
+		if userGroup == groupName {
+			return true
+		}
+	}
+
+	return false
+}
```

Comment:
Wouldn't be more explicit to use userName instead of groupName ?
'cos, for me groupName is the name of the group.

Reply:
🤔 

I'm confused haha. We're looking into a specific group name  in a list of groups the user is member of. Not a username 😅 

Though I agree, I'll rework.

Reply:
I've introduced a generic implementation of `Contains`, because that's what this function is actually doing 😅